instrumentation/spring/spring-integration-4.1/library/src/test/groovy/SpringCloudStreamRabbitTest.groovy

```diff
@@ -0,0 +1,131 @@
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration
+import org.springframework.cloud.stream.annotation.EnableBinding
+import org.springframework.cloud.stream.annotation.StreamListener
+import org.springframework.cloud.stream.messaging.Sink
+import org.springframework.cloud.stream.messaging.Source
+import org.springframework.context.ConfigurableApplicationContext
+import org.springframework.context.annotation.Bean
+import org.springframework.messaging.support.MessageBuilder
+import org.testcontainers.containers.GenericContainer
+import spock.lang.Shared
+
+class SpringCloudStreamRabbitTest extends LibraryInstrumentationSpecification {
+
+  @Shared
+  def rabbitMQContainer
+
+  @Shared
+  ConfigurableApplicationContext producerContext
+  @Shared
+  ConfigurableApplicationContext consumerContext
+
+  def setupSpec() {
+    rabbitMQContainer = new GenericContainer('rabbitmq:latest')
+      .withExposedPorts(5672)
+      .withStartupTimeout(Duration.ofSeconds(120))
+    rabbitMQContainer.start()
+
+    def producerApp = new SpringApplication(ProducerConfig, GlobalInterceptorSpringConfig)
+    producerApp.setDefaultProperties([
+      "spring.application.name"                        : "testProducer",
+      "spring.jmx.enabled"                             : false,
+      "spring.main.web-application-type"               : "none",
+      "spring.rabbitmq.host"                           : rabbitMQContainer.containerIpAddress,
+      "spring.rabbitmq.port"                           : rabbitMQContainer.getMappedPort(5672),
+      "spring.cloud.stream.bindings.output.destination": "testTopic"
+    ])
+    producerContext = producerApp.run()
+
+    def consumerApp = new SpringApplication(ConsumerConfig, GlobalInterceptorSpringConfig)
+    consumerApp.setDefaultProperties([
+      "spring.application.name"                       : "testConsumer",
+      "spring.jmx.enabled"                            : false,
+      "spring.main.web-application-type"              : "none",
+      "spring.rabbitmq.host"                          : rabbitMQContainer.containerIpAddress,
+      "spring.rabbitmq.port"                          : rabbitMQContainer.getMappedPort(5672),
+      "spring.cloud.stream.bindings.input.destination": "testTopic"
+    ])
+    consumerContext = consumerApp.run()
+  }
+
+  def cleanupSpec() {
+    rabbitMQContainer?.stop()
+    producerContext?.close()
+    consumerContext?.close()
+  }
+
+  def "should propagate context through RabbitMQ"() {
+    when:
+    producerContext.getBean("producer", Runnable).run()
+    consumerContext.getBean("latch", CountDownLatch).await(10, TimeUnit.SECONDS)
+
+    then:
+    assertTraces(1) {
+      trace(0, 4) {
+        span(0) {
+          name "producer"
+        }
+        span(1) {
+          name "testProducer.output"
+          childOf span(0)
+        }
+        span(2) {
+          name "testConsumer.input"
+          childOf span(1)
+        }
```

Comment:
sorry I forget, is the plan for these spans to have messaging semantic attributes and suppress the underlying rabbitmq spans?

Reply:
No, these were supposed to just propagate the context and coexist with actual underlying messaging instrumentations.
Though, now that I think of it, the reason why I added these spans was storing links; since we've decided on always getting the parent from the message they might not be necessary - the only thing that justifies their existence is adding custom `AttributesExtractor`s in the library instrumentation... I'll try removing the span in a separate PR.

Also, I'm planning to add a test that verifies how this instrumentation works with the rabbitmq instrumentation in the next PR: javaagent instrumentation.